tiledb/sm/filesystem/s3.cc

```diff
@@ -1332,37 +1345,66 @@ Status S3::write_multipart(
   const std::string uri_path(aws_uri.GetPath().c_str());
 
   // Take a lock protecting the shared multipart data structures
-  std::unique_lock<std::mutex> multipart_lck(multipart_upload_mtx_);
+  // Read lock to see if it exists
+  multipart_upload_mtx_.read_lock();
 
   auto state_iter = multipart_upload_states_.find(uri_path);
   if (state_iter == multipart_upload_states_.end()) {
-    // Delete file if it exists (overwrite) and initiate multipart request
-    bool exists;
-    RETURN_NOT_OK(is_object(uri, &exists));
-    if (exists) {
-      RETURN_NOT_OK(remove_object(uri));
-    }
+    // If the state is new, we must grab write lock, so unlock from read and
+    // grab write
+    multipart_upload_mtx_.read_unlock();
+    multipart_upload_mtx_.write_lock();
 
-    const Status st = initiate_multipart_request(aws_uri);
-    if (!st.ok()) {
-      return st;
-    }
+    // Since we switched locks we need to once again check to make sure another
+    // thread didn't create the state
+    auto state_iter = multipart_upload_states_.find(uri_path);
+    if (state_iter == multipart_upload_states_.end()) {
+      auto& path = aws_uri.GetPath();
+      std::string path_c_str = path.c_str();
+      MultiPartUploadState new_state;
+
+      assert(multipart_upload_states_.count(path_c_str) == 0);
+      multipart_upload_states_.emplace(
+          std::move(path_c_str), std::move(new_state));
+      MultiPartUploadState* state = &multipart_upload_states_.at(uri_path);
+      std::unique_lock<std::mutex> state_lck(state->mtx);
+      multipart_upload_mtx_.write_unlock();
+
+      // Delete file if it exists (overwrite) and initiate multipart request
+      bool exists;
+      RETURN_NOT_OK(is_object(uri, &exists));
+      if (exists) {
+        RETURN_NOT_OK(remove_object(uri));
+      }
 
-    state_iter = multipart_upload_states_.find(uri_path);
-    assert(state_iter != multipart_upload_states_.end());
+      const Status st = initiate_multipart_request(aws_uri, state);
+      if (!st.ok()) {
+        return st;
+      }
+      // Grab read lock, expected below outside the create
+      multipart_upload_mtx_.read_lock();
```

Comment:
The read unlock happens both here and L1388, we can make this a single `multipart_upload_mtx_.read_lock()` immediately after L1391